dmd, dtools: 2.106.1 -> 2.108.0, refactor

jtbx opened the pull request:
## Description of changes

https://dlang.org/changelog/2.107.0.html

**(edit 2024/03/04)** https://dlang.org/changelog/2.107.1.html
**(edit 2024/04/02)** https://dlang.org/changelog/2.108.0.html

Also migrate dmd and dtools to pkgs/by-name.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jtbx committed:
dmd: migrate to pkgs/by-name

jtbx commented:
Result of `nixpkgs-review pr 287854` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>dmd</li>
    <li>dtools</li>
  </ul>
</details>

jtbx commented:
Result of `nixpkgs-review pr 287854` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>dmd</li>
    <li>dtools</li>
  </ul>
</details>

kilianar reviewed (approved):
Result of `nixpkgs-review pr 287854` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>dmd</li>
    <li>dtools</li>
  </ul>
</details>

jtbx commented:
Marking this as draft since D 2.108.0 is almost here.

jtbx committed:
dmd: modernize

jtbx committed:
dtools: refactor

 o  switch to using finalAttrs pattern over rec expression in
    mkDerivation call
 o  use hash over sha256 in calls to fetchers
 o  remove the build-time dependency on gnumake42; dtools now builds
    fine with the latest version of GNU make
 o  use stdenv's implicit phases that build+install with make
 o  specify make flags through makeFlags rather than through an
    arbitrary attribute
 o  remove unnecessary make flags
 o  build with parallelism
 o  use checkTarget instead of providing a full checkPhase
 o  clean up meta

jtbx committed:
dtools: migrate to pkgs/by-name

jtbx committed:
dtools: 2.106.1 -> 2.108.0

Current arm64 patch broke so I adapted it to apply on 2.108.1.

The generic Makefile is now used instead of posix.mak which has been
deprecated for quite a while now.

jtbx committed:
dmd: 2.106.1 -> 2.108.0

Tests now use a D program called run.d, rather than the previous
Makefiles.

The compiler makefiles located in compiler/src/ have been
deprecated for a long time now and have finally been removed in favour
of build.d. There is also a top-level generic Makefile which also
works on Windows apparently. [1]

The Phobos makefiles (posix.mak and such) have also been removed and
replaced with a generic top-level Makefile. Same with druntime. [2]

[1]: <https://dlang.org/changelog/2.107.0.html#dmd.makefiles>
[2]: <https://dlang.org/changelog/2.107.0.html#druntime.makefiles>
     <https://dlang.org/changelog/2.107.0.html#makefiles>

jtbx commented:
Updated to 2.108.0

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3739

kilianar commented:
Result of `nixpkgs-review pr 287854` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>dmd</li>
    <li>dtools</li>
  </ul>
</details>

lucasnethaj commented:
Result of `nixpkgs-review pr 287854` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>dmd</li>
    <li>dtools</li>
  </ul>
</details>

imrying commented:
Result of `nixpkgs-review pr 287854` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>dmd</li>
    <li>dtools</li>
  </ul>
</details>

jtbx commented:
Thanks! <3
